rules/testutil/testutil.go

```diff
@@ -13,95 +13,119 @@ import (
-func descriptorProtoFromSource(source io.Reader) (*descriptorpb.FileDescriptorProto, error) {
+// MustCreateFileDescriptorProtoFromTemplate creates a *descriptorpb.FileDescriptorProto from a string template and data.
+func MustCreateFileDescriptorProtoFromTemplate(filename, srcTmpl string, data interface{}, deps []*descriptorpb.FileDescriptorProto) *descriptorpb.FileDescriptorProto {
+	tmpl := template.Must(template.New("test").Parse(srcTmpl))
+	b := new(bytes.Buffer)
+	if err := tmpl.Execute(b, data); err != nil {
+		log.Fatalf("Error executing template %v", err)
+	}
+
+	return mustCreateDescriptorProtoFromSource(filename, b, deps)
+}
+
+func mustCreateDescriptorProtoFromSource(filename string, source io.Reader, deps []*descriptorpb.FileDescriptorProto) *descriptorpb.FileDescriptorProto {
 	tmpDir := os.TempDir()
 
 	f, err := ioutil.TempFile(tmpDir, "proto*")
 	if err != nil {
-		return nil, err
+		log.Fatalf("Failed creating temp proto source file: %s", err)
 	}
 	defer mustCloseAndRemoveFile(f)
 
 	if _, err = io.Copy(f, source); err != nil {
-		return nil, err
+		log.Fatalf("Failed to copy source to templ file: %s", err)
 	}
 
 	descSetF, err := ioutil.TempFile(tmpDir, "descset*")
 	if err != nil {
-		return nil, err
+		log.Fatalf("Failed to create temp descriptor set file: %s", err)
 	}
 	defer mustCloseAndRemoveFile(descSetF)
 
-	cmd := exec.Command(
-		protocPath(),
+	args := []string{
 		"--include_source_info",
 		fmt.Sprintf("--proto_path=%s", tmpDir),
 		fmt.Sprintf("--descriptor_set_out=%s", descSetF.Name()),
-		f.Name(),
-	)
+	}
+
+	if len(deps) > 0 {
+		descSetIn := mustCreateDescSetFileFromFileDescriptorProtos(deps)
+		defer mustCloseAndRemoveFile(descSetIn)
+
+		args = append(args, fmt.Sprintf("--descriptor_set_in=%s", descSetIn.Name()))
+	}
+
+	args = append(args, f.Name())
+
+	cmd := exec.Command(protocPath(), args...)
 
 	stderr := new(bytes.Buffer)
 	cmd.Stderr = stderr
 
 	if err = cmd.Run(); err != nil {
-		return nil, fmt.Errorf("protoc failed with %v and Stderr %q", err, stderr.String())
+		log.Fatalf("protoc failed with %v and Stderr %q", err, stderr.String())
 	}
 
 	descSet, err := ioutil.ReadFile(descSetF.Name())
 	if err != nil {
-		return nil, err
+		log.Fatalf("Failed to read descriptor set file: %s", err)
 	}
 
 	protoset := &descriptorpb.FileDescriptorSet{}
 	if err := proto.Unmarshal(descSet, protoset); err != nil {
-		return nil, err
+		log.Fatalf("Failed to unmarshal descriptor set file: %s", err)
 	}
 
 	if len(protoset.GetFile()) == 0 {
-		return nil, fmt.Errorf("protoset.GetFile() returns empty list")
+		log.Fatalf("protoset.GetFile() returns empty list")
 	}
 
-	return protoset.GetFile()[0], nil
+	protoset.GetFile()[0].Name = &filename
+
+	return protoset.GetFile()[0]
 }
 
-func mustCloseAndRemoveFile(f *os.File) {
-	if err := f.Close(); err != nil {
-		log.Fatalf("Error closing proto file: %v", err)
+func mustCreateDescSetFileFromFileDescriptorProtos(descs []*descriptorpb.FileDescriptorProto) *os.File {
```

Comment:
Let's rewrite this function to `writeFileDescriptorSet(filename string, s *descriptorpb.FileDescriptorSet) error`, so the caller can create the temp file in a consistent place and remember to remove it.

Reply:
Any particular reason for this preference? I think it's a bit cleaner for the caller to not have to deal with the temp file as well as handling two extra errors (one from ioutil.TempFile and one from this function), but it really just seems like a matter of preference. Also, this is an unexported method with only one usage, so this is an extremely low impact decision.